internal/pkg/deploy/cloudformation/bb_pipeline_integration_test.go

```diff
@@ -0,0 +1,182 @@
+			AccountID: appCallerInfo.Account,
+		}
+		pipelineStackName := app.Name + "-pipepiper"
+		appRoleStackName := fmt.Sprintf("%s-infrastructure-roles", app.Name)
+		appStackSetName := fmt.Sprintf("%s-infrastructure", app.Name)
+
+		// find another region (different from the application region,
+		// i.e. *sess.Config.Region) for us to deploy an environment in.
+		envRegion, err := findUnusedRegion("us-west", *appSess.Config.Region)
+		require.NoError(t, err)
+		envSess, err := testSession(aws.String(envRegion.ID()))
+		require.NoError(t, err)
+		envCfClient := awsCF.New(envSess)
+		envId := identity.New(envSess)
+		envCallerInfo, err := envId.Get()
+		require.NoError(t, err)
+		envDeployer := cloudformation.New(envSess)
+
+		environmentToDeploy := deploy.CreateEnvironmentInput{
+			Name:                     randStringBytes(10),
+			AppName:                  app.Name,
+			ToolsAccountPrincipalARN: envCallerInfo.RootUserARN,
+		}
+		envStackName := fmt.Sprintf("%s-%s",
+			environmentToDeploy.AppName,
+			environmentToDeploy.Name)
+
+		// Make sure we delete the stacks after the test is done
+		defer func() {
+			// delete the pipeline first because it relies on stackset
+			_, err := appCfClient.DeleteStack(&awsCF.DeleteStackInput{
+				StackName: aws.String(pipelineStackName),
+			})
+			require.NoError(t, err)
+			err = appCfClient.WaitUntilStackDeleteComplete(&awsCF.DescribeStacksInput{
+				StackName: aws.String(pipelineStackName),
+			})
+			require.NoError(t, err)
+
+			// Clean up any StackInstances we may have created.
+			if stackInstances, err := appCfClient.ListStackInstances(&awsCF.ListStackInstancesInput{
+				StackSetName: aws.String(appStackSetName),
+			}); err == nil && stackInstances.Summaries != nil && stackInstances.Summaries[0] != nil {
+				appStackInstance := stackInstances.Summaries[0]
+				_, err := appCfClient.DeleteStackInstances(&awsCF.DeleteStackInstancesInput{
+					Accounts:     []*string{appStackInstance.Account},
+					Regions:      []*string{appStackInstance.Region},
+					RetainStacks: aws.Bool(false),
+					StackSetName: appStackInstance.StackSetId,
+				})
+				require.NoError(t, err)
+
+				err = appCfClient.WaitUntilStackDeleteComplete(&awsCF.DescribeStacksInput{
+					StackName: appStackInstance.StackId,
+				})
+				require.NoError(t, err)
+			}
+			// Delete the StackSet once all the StackInstances are cleaned up
+			_, err = appCfClient.DeleteStackSet(&awsCF.DeleteStackSetInput{
+				StackSetName: aws.String(appStackSetName),
+			})
+			require.NoError(t, err)
+
+			_, err = appCfClient.DeleteStack(&awsCF.DeleteStackInput{
+				StackName: aws.String(appRoleStackName),
+			})
+			require.NoError(t, err)
+			err = appCfClient.WaitUntilStackDeleteComplete(&awsCF.DescribeStacksInput{
+				StackName: aws.String(appRoleStackName),
+			})
+			require.NoError(t, err)
+
+			// delete the environment stack once we are done
+			_, err = envCfClient.DeleteStack(&awsCF.DeleteStackInput{
```

Comment:
@bvtujo mentioned that we are almost about to hit the number of IAM role limits in our test accounts :D this is because after deleting the stack we don't delete the IAM roles that were retained:
https://github.com/aws/copilot-cli/blob/12eeeadce76648656e861437751309c043df7e64/templates/environment/partials/environment-manager-role.yml#L1-L4

https://github.com/aws/copilot-cli/blob/12eeeadce76648656e861437751309c043df7e64/templates/environment/partials/cfn-execution-role.yml#L3-L6

Maybe not in this PR but it would be awesome if on the `mainline` branch we can update the integ tests to delete these roles after the stack is deleted.

Reply:
Ah. I'll make a note.